### ISSUE TITLE
[mso_schema_site] Add full test file and fix delete bug

### ISSUE THREAD

**Issue by lhercot** [CiscoDevNet/ansible-mso#91]
When trying to delete a site this error happens:

`failed: [localhost] (item=TEST_SITE) => {"ansible_loop_var": "item", "changed": true, "current": {}, "data": [{"op": "remove", "path": "/sites/TEST_SITE"}], "info": {"body": "{\"code\":500,\"message\":\"Template name should be provided like siteId-TemplateName\"}", "connection": "close", "content-language": "en", "content-length": "82", "content-type": "application/json", "date": "Thu, 06 Aug 2020 00:23:58 GMT", "msg": "HTTP Error 500: Internal Server Error", "server": "nginx", "status": 500, "strict-transport-security": "max-age=31536000; includeSubDomains", "url": "https://10.66.176.111/api/v1/schemas/5f2b47699700006e02d10eb8", "x-ratelimit-limit": "-1", "x-ratelimit-remaining": "-1", "x-ratelimit-reset": "1559582945"}, "item": "TEST_SITE", "msg": "MSO Error 500: Template name should be provided like siteId-TemplateName", "payload": {"code": 500, "message": "Template name should be provided like siteId-TemplateName"}}`

**Comment by kentwu1** [comment]
I also can't delete the template thru mso_schema_template module. And here is the error msg I ran into,

failed: [localhost] (item=Fab2) => {"ansible_loop_var": "item", "changed": true, "current": {}, "data": [{"op": "remove", "path": "/templates/Fab2"}], "info": {"body": "{\"code\":400,\"message\":\"Bad Request: Template 'Fab2' referenced in a site local entry is not defined on schema 'kent_test1'\"}", "connection": "close", "content-language": "en", "content-length": "124", "content-type": "application/json", "date": "Thu, 06 Aug 2020 00:07:56 GMT", "msg": "HTTP Error 400: Bad Request", "server": "nginx", "status": 400, "strict-transport-security": "max-age=31536000; includeSubDomains", "url": "https://10.66.176.111/api/v1/schemas/5f2b47699700006e02d10eb8", "x-ratelimit-limit": "-1", "x-ratelimit-remaining": "-1", "x-ratelimit-reset": "1559582945"}, "item": "Fab2", "msg": "MSO Error 400: Bad Request: Template 'Fab2' referenced in a site local entry is not defined on schema 'kent_test1'", "payload": {"code": 400, "message": "Bad Request: Template 'Fab2' referenced in a site local entry is not defined on schema 'kent_test1'"}}

**Comment by shrsr** [comment]
Opened a new issue for deletion of template using mso_schema_template

**Comment by shrsr** [comment]
@kentwu1 -
In order to delete the template using mso_schema_template we have to first remove the site, the template is associated with.
We can use mso_schema_site to achieve this.
Please use this as an example:

- name: Remove site from Template 1 of Schema1
  mso_schema_site:
    <<: *mso_info
    schema: Schema 1
    site: Site 1
    template: Template 1
    state: absent

- name: Remove Template 1 of Schema 1
  mso_schema_template:
    <<: *mso_info
    schema: Schema 1
    tenant: ansible_test
    template: Template 1
    state: absent

**Comment by kentwu1** [comment]
ok will try. Thx.

**Comment by shrsr** [comment]
Btw a PR is pending to be merged. What I said will work once that PR is merged. That should happen soon.
Will let you know as soon as it is done.

-Thanks

**Comment by shrsr** [comment]
@kentwu1 
It's done. You can try it out now.

**Comment by kentwu1** [comment]
So how do I get those newly released MSO modules? Do I have to upgrade my ansible then? Thx.

**Comment by kentwu1** [comment]
Tried to verify it but looks like I can't just replace the mso_schema_site.py file directly so pls advise. Here is the error I ran into,

TASK [Remove site2 from templates] *****************************************************************************
fatal: [localhost]: FAILED! => {"msg": "Could not find imported module support code for mso_schema_site.  Looked for either MSOModule.py or mso.py"}

And below is the diff between new and the old version installed on my system,

noiro@k8s21-node-1:/usr/lib/python2.7/dist-packages/ansible/modules/network/aci$ git diff mso_schema_site.py mso_schema_site.py.orig
diff --git a/mso_schema_site.py b/mso_schema_site.py.orig
index 101068b..751262f 100644
--- a/mso_schema_site.py
+++ b/mso_schema_site.py.orig
@@ -2,8 +2,7 @@
 # -*- coding: utf-8 -*-
 
 # Copyright: (c) 2018, Dag Wieers (@dagwieers) <dag@wieers.com>
-# Copyright: (c) 2020, Shreyas Srish (@shrsr) <ssrish@cisco.com>
-# GNU General Public License v3.0+ (see LICENSE or https://www.gnu.org/licenses/gpl-3.0.txt)
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
 
 from __future__ import absolute_import, division, print_function
 __metaclass__ = type
@@ -20,7 +19,7 @@ description:
 - Manage sites on Cisco ACI Multi-Site.
 author:
 - Dag Wieers (@dagwieers)
-- Shreyas Srish (@shrsr)
+version_added: '2.8'
 options:
   schema:
     description:
@@ -31,10 +30,12 @@ options:
     description:
     - The name of the site to manage.
     type: str
+    required: yes
   template:
     description:
     - The name of the template.
     type: str
+    required: yes
     aliases: [ name ]
   state:
     description:
@@ -44,52 +45,53 @@ options:
     choices: [ absent, present, query ]
     default: present
 seealso:
-- module: cisco.mso.mso_schema_template
-- module: cisco.mso.mso_site
-extends_documentation_fragment: cisco.mso.modules
+- module: mso_schema_template
+- module: mso_site
+extends_documentation_fragment: mso
 '''
 
 EXAMPLES = r'''
 - name: Add a new site to a schema
-  cisco.mso.mso_schema_site:
+  mso_schema_site:
     host: mso_host
     username: admin
     password: SomeSecretPassword
     schema: Schema 1
-    site: Site1
+    site: bdsol-pod51
     template: Template 1
     state: present
   delegate_to: localhost
 
 - name: Remove a site from a schema
-  cisco.mso.mso_schema_site:
+  mso_schema_site:
     host: mso_host
     username: admin
     password: SomeSecretPassword
     schema: Schema 1
-    site: Site1
+    site: bdsol-pod51
     template: Template 1
     state: absent
   delegate_to: localhost
 
 - name: Query a schema site
-  cisco.mso.mso_schema_site:
+  mso_schema_site:
     host: mso_host
     username: admin
     password: SomeSecretPassword
     schema: Schema 1
-    site: Site1
+    site: bdsol-pod51
     template: Template 1
     state: query
   delegate_to: localhost
   register: query_result
 
 - name: Query all schema sites
-  cisco.mso.mso_schema_site:
+  mso_schema_site:
     host: mso_host
     username: admin
     password: SomeSecretPassword
     schema: Schema 1
+    site: bdsol-pod51
     state: query
   delegate_to: localhost
   register: query_result
@@ -99,7 +101,7 @@ RETURN = r'''
 '''
 
 from ansible.module_utils.basic import AnsibleModule
-from ansible_collections.cisco.mso.plugins.module_utils.mso import MSOModule, mso_argument_spec
+from ansible.module_utils.network.aci.mso import MSOModule, mso_argument_spec
 
 
 def main():
@@ -120,10 +122,10 @@ def main():
         ],
     )
 
-    schema = module.params.get('schema')
-    site = module.params.get('site')
-    template = module.params.get('template')
-    state = module.params.get('state')
+    schema = module.params['schema']
+    site = module.params['site']
+    template = module.params['template']
+    state = module.params['state']
 
     mso = MSOModule(module)
 
@@ -140,14 +142,13 @@ def main():
 
     mso.existing = {}
     if 'sites' in schema_obj:
-        sites = [(s.get('siteId'), s.get('templateName')) for s in schema_obj.get('sites')]
+        sites = [(s['siteId'], s['templateName']) for s in schema_obj['sites']]
         if template:
             if (site_id, template) in sites:
                 site_idx = sites.index((site_id, template))
-                site_path = '/sites/{0}'.format(site_idx)
-                mso.existing = schema_obj.get('sites')[site_idx]
+                mso.existing = schema_obj['sites'][site_idx]
         else:
-            mso.existing = schema_obj.get('sites')
+            mso.existing = schema_obj['sites']
 
     if state == 'query':
         if not mso.existing:
@@ -158,6 +159,7 @@ def main():
         mso.exit_json()
 
     sites_path = '/sites'
+    site_path = '/sites/{0}'.format(site)
     ops = []
 
     mso.previous = mso.existing

**Comment by shrsr** [comment]
@kentwu1 
One way of updating the mso on your system is:-
1. Go to: https://github.com/CiscoDevNet/ansible-mso/actions
2. Select the latest CI build
3. Under Artifacts download collections and unzip it on Terminal
4. Get the tar.gz file
5. Install using ansible-galaxy collection install 
6. Example: ansible-galaxy collection install cisco-mso-1.0.0.tar.gz —force

Run your playbook once this is done.

**Comment by kentwu1** [comment]
ok it does work as expected after following your instruction. Thx!

**Comment by kentwu1** [comment]
One related issue I ran into now is when I tried to remove a site from a template, its complaining I have to undeploy that template first even though the undeploy was already done successfully earlier....If I ran the [Remove site2 from templates] task again then it will go thru this time. So looks like there might be a delay for MSO to reflect the right status after a template has been undeployed even though ansible has returned ok? And below is the related tasks execution sequence & error msg. Let me know if I should open a new issue for this then. Thx.

TASK [Undeploy schema templates] *******************************************************************************
ok: [localhost] => (item=Common)
ok: [localhost] => (item=Fab1)

TASK [Undeploy schema templates] *******************************************************************************
ok: [localhost] => (item=Common)
ok: [localhost] => (item=Fab2)   <=== undeploy of Fab2 is already done here.

TASK [Remove site1 from templates] *****************************************************************************
changed: [localhost] => (item=Common)
changed: [localhost] => (item=Fab1)

TASK [Remove site2 from templates] *****************************************************************************
changed: [localhost] => (item=Common)
failed: [localhost] (item=Fab2) => {"ansible_loop_var": "item", "changed": true, "current": {}, "data": [{"op": "remove", "path": "/sites/0"}], "info": {"body": "{\"code\":400,\"message\":\"Bad Request: Site 5f07e0bd960000410018cb70 cannot be deleted; template Fab2 must first be undeployed\"}", "connection": "close", "content-language": "en", "content-length": "125", "content-type": "application/json", "date": "Tue, 25 Aug 2020 23:19:58 GMT", "msg": "HTTP Error 400: Bad Request", "server": "nginx", "status": 400, "strict-transport-security": "max-age=31536000; includeSubDomains", "url": "https://10.66.176.111/api/v1/schemas/5f459b3698000032cd718109", "x-ratelimit-limit": "-1", "x-ratelimit-remaining": "-1", "x-ratelimit-reset": "1559582945"}, "item": "Fab2", "msg": "MSO Error 400: Bad Request: Site 5f07e0bd960000410018cb70 cannot be deleted; template Fab2 must first be undeployed", "payload": {"code": 400, "message": "Bad Request: Site 5f07e0bd960000410018cb70 cannot be deleted; template Fab2 must first be undeployed"}}

**Comment by shrsr** [comment]
Can you please open another issue for this? 
Please share the playbook you used for the above on the new issue.

**Comment by kentwu1** [comment]
ok will do.

**Comment by kentwu1** [comment]
I ran my playbook 3 times today but still couldn't reproduce this anymore. Not sure why it was pretty consistent yesterday.....I will just keep an eye on it for now then.
